Fix bug with renaming columns on import

seancolsen opened the pull request:
Fixes #4372

## About the underlying problem

In one of the many steps we perform when confirming an import, the front end hits `columns.patch` in order to update the column names and types. Prior to this PR, it sent `name`, `type`, and `type_options` for every column in the table. So in Mathesar 0.2.1 even for the auto-generated PK column (which we don't allow the user to cast), the front end was still saying:

> yes DB-layer, please change the type of the `id` column to "integer"

The backend was able to handle this just fine. It should be a no-op after all.

But for some reason the backend is _not_ able to gracefully handle that same sort of request when made against columns that originated from the CSV and were subsequently configured to have IDENTITY defaults. (Interestingly, I noticed that if you choose a pre-existing numeric column and take the extra step of changing the default to None, then this bug is not present.)

I haven't dug into the backend code, but just from looking at the error message, I can spot this:

```sql
ALTER TABLE "Table 7"
ALTER COLUMN a DROP DEFAULT,
ALTER COLUMN a TYPE integer USING msar.cast_to_integer(a),
ALTER COLUMN a SET DEFAULT ::integer,
--                         ^^^^^^^^^ 👈 Here is the syntax error
ALTER COLUMN name DROP DEFAULT,
ALTER COLUMN name TYPE text USING msar.cast_to_text(name)
```

Perhaps one could argue this is a backend bug. I don't know.

## About the fix

In the end I chose to fix this on the front end by just never sending `type` and `type_options` _for PK columns_. The front end does not allow the user to change those settings for PK columns, and I think it ought to stay that way.


## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

seancolsen commented:
@pavish I'd like you to review front end code.

@mathemancer I'd like you to read over my assessment of the problem and solution to make sure it seems sane to you.

zackkrida reviewed (approved):
@seancolsen your code makes sense and the comments are useful. I see that column `metadata` is no longer being returned for _any_ column regardless of whether it is a primary key. Are you certain that wasn't being used for anything? If so, LGTM.

seancolsen commented:
> I see that column metadata is no longer being returned for any column regardless of whether it is a primary key. Are you certain that wasn't being used for anything?

My, what a keen eye you have, Zack!

Yes I'm certain that removing `metadata` is safe here. Here's why:

1. In the back end, we can look at the [columns.add RPC handler function](https://github.com/mathesar-foundation/mathesar/blob/3c095f9b1cf7e85c859bf1fc424c22e756f00198/mathesar/rpc/columns/base.py#L282). 
    - It's [SettableColumnInfo](https://github.com/mathesar-foundation/mathesar/blob/3c095f9b1cf7e85c859bf1fc424c22e756f00198/mathesar/rpc/columns/base.py#L118) parameter type does not have a `metadata` property.
    - But being that we current do not use tooling to enforce types on the back end, we can quickly trace this function body down too, and see that `column_data_list` only every gets passed to the DB layer. So there is effectively no way it would be using an undocumented `metadata` property. 

1. On the front end, this means that our [ColumnPatchSpec](https://github.com/mathesar-foundation/mathesar/blob/3c095f9b1cf7e85c859bf1fc424c22e756f00198/mathesar_ui/src/api/rpc/columns.ts#L191) type is correct — it does not have a `metadata` property. Because we _do_ use a type-checker on the front end, we can be quite confident we're safe! Yay TS! 🙂

---

Read on, only if you're curious...

Ok, so then why was `metadata` there?

- It was likely left over from and old REST API that _did_ support metadata changes in the same API as column type/name modification. Now we handle this sort of combination via RPC batching.

Ok, so then why was there no TS error _before_ my change?

- I actually removed metadata because TS complained at me after my changes. Gosh I ♥ TS!

- It's kind of fun to see why there was no TS error before....
    
- Prior to my change, [finalizeColumns](https://github.com/mathesar-foundation/mathesar/blob/3c095f9b1cf7e85c859bf1fc424c22e756f00198/mathesar_ui/src/pages/import/preview/importPreviewPageUtils.ts#L117) was returning an array of `ColumnPatchSpec` _subtypes_ — objects that conformed to the `ColumnPatchSpec` interface but had an extra "metadata" property too. That subtype got fed into [updateTable](https://github.com/mathesar-foundation/mathesar/blob/3c095f9b1cf7e85c859bf1fc424c22e756f00198/mathesar_ui/src/stores/tables.ts#L224). Because type substitution is contravariant in function parameters, there was no TS error here. A subtype can be substituted for a supertype as a function argument.
    
- After my change, I added `finalizeColumn`. In doing so, I added an explicit return type. (I could have left this to inference, but it seemed like a good idea here.) But function return types are _covariant_! So I can't substitute a subtype in place of a supertype. I had to make the return value conform to `ColumnPatchSpec` _exactly_.

Anyway, sorry for the long winded comment. Honestly, writing this was an opportunity for me to shore up my understanding of variance. I only just began learning about this after seeing the TS variance annotations feature added to TS somewhat recently.

zackkrida commented:
Thanks @seancolsen. That was an enjoyable read.

mathemancer reviewed (commented):
I agree we should also fix this behavior so that sending type details is a no-op in the back end.
